upload/admin/controller/smailyforopencart/upgrade.php

```diff
@@ -66,5 +66,22 @@ public function upgrade($route, $args, $output) {
                 }
             }
         }
+
+        // Version 1.3.2 - new event smaily_reset_empty_cart.
+        if (version_compare($version, '1.3.2', '>=')) {
```

Comment:
The upgrade process is not so good for OpenCart. You can basically check only the current version of installed module because OpenCart won't let you install modules with the same code. You have to uninstall previous version from the modification table and then add a new one. So the `>=` version compare compares if a newer version of the plugin is being installed after current change.

Reply:
Hm, but shouldn't `uninstall` and `install` method be executed then? I mean why is this upgrade function even necessary.

Reply:
There is no official "upgrade" logic/hooks that I have found of. The current implementation is so we can provide some kind of upgrade scripts with our plugin. OpenCart module installation and "extension" installation are considered separate logic. The idea to me is kind of funky.
- You can install an extension with a `zip` file and that is just uploading content to the root of the OpenCart folder.
- You can install a module by clicking the install link on the module page. This now runs the install script and installs table, event hooks, and whatnot.
- You can then remove the extension from the extension table to install a new version of our plugin. This just removes the entry from the `modification` table and does not remove files. The old module is still installed and active.
- Now you can then overwrite old module files by installing a new module and creating a new entry to the `modification` table by uploading a new version with the extension installer. This upgrade script runs.
- The uninstall function does not remove files. It basically reverses the instructions in the `install.xml` file that modify some kind of core files directly. Our plugin doesn't do that.

Reply:
> I mean why is this upgrade function even necessary.

The user doesn't have to install our plugin again or even activate it again after the upgrade. So the only way to update hooks/tables is to run the upgrade script check when upgrading our plugin. There is an issue when files are just copied. Then there is no way to control this kind of behavior.

Reply:
OK